nodego/events.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodego
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"google.golang.org/api/pubsub/v1"
+	"google.golang.org/api/storage/v1"
+)
+
+// JSONTime is a wrapper for time.Time to decode the time from json.
+type JSONTime struct {
+	time.Time
+}
+
+// UnmarshalJSON parses a json string to time.Time.
+func (t *JSONTime) UnmarshalJSON(b []byte) (err error) {
+	if string(b) != "null" {
+		t.Time, err = time.Parse(`"2006-01-02T15:04:05.000Z"`, string(b))
+	}
+	return
+}
+
+// FunctionEvent is the basic data structure passed to functions by non-http
+// triggers.
+type FunctionEvent struct {
+	EventID   string          `json:"eventId"`
+	Timestamp JSONTime        `json:"timestamp"`
+	EventType string          `json:"eventType"`
+	Resource  string          `json:"resource"`
+	Data      json.RawMessage `json:"data"`
+}
+
+// PubSubMessage is a wrapper for pubsub.PubsubMessage
+type PubSubMessage struct {
+	pubsub.PubsubMessage
+
+	Data []byte
+}
+
+// PubSubMessage unmarshals the event data as a pub sub message.
+func (e *FunctionEvent) PubSubMessage() (*PubSubMessage, error) {
+	var msg pubsub.PubsubMessage
+	if err := json.Unmarshal(e.Data, &msg); err != nil {
+		return nil, err
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(msg.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PubSubMessage{
+		PubsubMessage: msg,
+		Data:          decoded,
+	}, nil
+}
+
+// StorageObject is a wrapper for storage.Object
+type StorageObject struct {
+	storage.Object
+
+	// TODO maybe add pre-parsed time fields
+}
+
+// StorageEvent unmarshals the event data as a storage event.
+func (e *FunctionEvent) StorageObject() (*StorageObject, error) {
+	var obj storage.Object
+	if err := json.Unmarshal(e.Data, &obj); err != nil {
+		return nil, err
+	}
+
+	return &StorageObject{
+		Object: obj,
+	}, nil
+}
+
+// EventHandler returns http.Handler that parses the body for a function event.
+func EventHandler(handler func(*FunctionEvent) error) http.Handler {
```

Comment:
I think if you return `http.HandlerFunc` this can be used with either `http.Handle` or `http.HandleFunc`.